### PR TITLE
[KYUUBI #5900] Support to deregister the engine on open failure

### LINKED ISSUE #5900: [FEATURE] Support to delete the engine node from ha namespace if failed to connect to engine host:port
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Support to delete the engine node from ha namespace if failed to connect to engine host:port

Now if this situation happen(for example, hive client initialization failed because of issue), we have to delete the engine manually.

### Motivation

_No response_

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by turboFei**
# :mag: Description
## Issue References 🔗

This pull request fixes #5900

## Describe Your Solution 🔧

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Related Unit Tests

---

# Checklists
## 📝 Author Self Checklist

- [ ] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [ ] I have performed a self-review
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [ ] Pull request title is okay.
- [ ] No license issues.
- [ ] Milestone correctly set?
- [ ] Test coverage is ok
- [ ] Assignees are selected.
- [ ] Minimum number of approvals
- [ ] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

**Review by wForget — APPROVED** [review]
LGTM

**Comment by wForget** [comment]
There seem to be related failures.

**Comment by wForget** [comment]
I think this may help us reset engine when the shared engine is stuck, can we plan to add it to branch-1.8?

**Comment by pan3793** [comment]
The "reset" is not an intuitive word for end users

**Comment by turboFei** [comment]
> The "reset" is not an intuitive word for end users

Do you have some suggestions?

**Comment by pan3793** [comment]
The behavior here may be more flexible, e.g. should we "reset" the engine immediately after the first failure, or after max attempts is reached? Should we kill the engine or simply delete the znode?

**Comment by turboFei** [comment]
@pan3793 how about now?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -1516,6 +1516,14 @@ object KyuubiConf {
       .timeConf
       .createWithDefault(Duration.ofSeconds(10).toMillis)
 
+  val ENGINE_OPEN_RESET_ON_FAILURE: ConfigEntry[Boolean] =
+    buildConf("kyuubi.session.engine.open.reset.onFailure")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1520]: maybe something like `kyuubi.session.engine.open.onFailure`

```
- RETRY
- DEREGISTER_IMMEDIATELY
- DEREGISTER_AFTER_RETRY
- KILL_IMMEDIATELY (leave option for future implementation)
- KILL_AFTER_RETRY (leave option for future implementation)
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -1516,6 +1516,22 @@ object KyuubiConf {
...
+    val RETRY, DEREGISTER_IMMEDIATELY, DEREGISTER_AFTER_RETRY = Value
+  }
+
+  val ENGINE_OPEN_ON_FAILURE: ConfigEntry[String] =
+    buildConf("kyuubi.session.engine.open.onFailure")
+      .doc("The behavior when opening engine failed. " +
+        "RETRY: retry to open engine for ENGINE_OPEN_MAX_ATTEMPTS times. " +
+        "DEREGISTER_IMMEDIATELY: deregister the session immediately. " +
+        "DEREGISTER_AFTER_RETRY: deregister the session after retry to open engine for " +
+        "ENGINE_OPEN_MAX_ATTEMPTS times.")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:1530]: replace ENGINE_OPEN_MAX_ATTEMPTS with the conf key

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 3]
```diff
@@ -307,6 +307,19 @@ private[kyuubi] class EngineRef(
       }
   }
 
+  /**
+   * Reset the engine ref from engine space with the given host and port on connection failure.
+   *
+   * @param discoveryClient the zookeeper client to get or create engine instance
+   * @param hostPort the exsiting engine host and port
+   */
+  def reset(discoveryClient: DiscoveryClient, hostPort: (String, Int)): Unit =
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:316]: deregister

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala [hunk 4]
```diff
@@ -307,6 +307,19 @@ private[kyuubi] class EngineRef(
       }
   }
 
+  /**
+   * Deregister the engine from engine space with the given host and port on connection failure.
+   *
+   * @param discoveryClient the zookeeper client to get or create engine instance
+   * @param hostPort the existing engine host and port
+   */
+  def deregister(discoveryClient: DiscoveryClient, hostPort: (String, Int)): Unit =
```
Comment (Kwafoor) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:316]: Look like this method can be used in #5871   to deregister engine?

Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala:316]: looks good.

**Comment by turboFei** [comment]
thanks, merged to master and 1.8.1

**Pull request closed without merging**